### ISSUE TITLE
Make all structs that claim to be thread-safe dereference `this` at most once per entrypoint

### ISSUE THREAD

**Issue by AArnott** [dotnet/coreclr#1167]
`ImmutableArray<T>` goes to great lengths to be thread-safe by [never dereferencing `this` more than once per public entrypoint][1] (to fix bug #191). Other structs in the Framework do *not* take this care (such as `DateTime`, `Nullable<T>`, or `System.Int32`), despite [claiming to be thread-safe in MSDN documentation][2].

Should we take the same care for the other structs? 

The community [has been fascinated][3] by the pattern in `ImmutableArray<T>` and brought up that it might make sense to spread it around more.

 [1]: https://github.com/dotnet/corefx/blob/1dc118cc46f88e4889a23190b450d89e565967fd/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray%601.cs#L18-L33
 [2]: https://msdn.microsoft.com/en-us/library/System.DateTime(v=vs.110).aspx
 [3]: http://stackoverflow.com/questions/30957211/what-advantage-is-there-to-storing-this-in-a-local-variable-in-a-struct-method/30957262

**Comment by hadibrais** [comment]
This is only practical if the size of the struct is equal to the size of a reference, which is not the case for **DateTime** and **Nullable&lt;T&gt;** and many other structs on x86. That's because some synchronization mechanism has to be used to copy the instance atomically which can add significant overhead.

If we make this change, all structs will be divided into two categories: those whose size is equal to a reference and those who aren't. If we want to apply this change to all types in the first category, I believe it would be better to modify the JIT compiler so that it automatically makes a copy of the instance on every method call. This eliminates any maintainability overhead associated with this issue and makes sure all methods of all such structs follow this pattern automatically.

**Comment by omariom** [comment]
In addition to atomicity @hadibrais mentioned:
If it is about thread safety take into account that in the future JITor even C# compiler (or other CLI JITs and C# compilers) can decide to eliminate reads when find it useful. So technically the read should be  volatile one. Though I don't know if it possible to volatile read *this* in structs' methods. And volatile reads on ARM are  sort of heavy.

**Comment by mikedn** [comment]
> and Nullable<T> and many other structs on x86

`Nullable<T>` isn't atomic on any platform, not only on x86. In general, any struct containing more than one field is not guaranteed to be atomic.

>  I believe it would be better to modify the JIT compiler so that it automatically makes a copy of the instance on every method call

That can introduce subtle changes in the behavior of existing code. Not to mention possible performance issues. The JIT compiler has enough problems with structs as it is.

> despite claiming to be thread-safe in MSDN documentation.

That's not a very accurate description of the situation: 
* `Nullable<T>` [doesn't claim to be thread safe](https://msdn.microsoft.com/en-us/library/b3h38hb0).
* `DateTime` claims to be thread safe but at the same time it warns that assignment is not an atomic operation and assignment is the only way to introduce thread safety issue in the case of DateTime. 
* `Int32` claims thread-safety but with a note about modifying Int32 variables: "All members of this type are thread safe. Members that appear to modify instance state actually return a new instance initialized with the new value. **As with any other type, reading and writing to a shared variable that contains an instance of this type must be protected by a lock to guarantee thread safety**."

I suppose it would be nice if [Int32.CompareTo](http://referencesource.microsoft.com/#mscorlib/system/int32.cs,67) would read `m_value` only once but if you're going to call `CompareTo` and modify the `int` at the same time without any synchronization then you're going to get burned in many cases no matter what `CompareTo` does.

In summary I think that this is primarily a problem with the documentation's use of the "thread safe" term as a catch all bucket for all thread related issues and some reader's unfamiliarity with all the subtleties involved which leads to invalid expectations and assumptions.

**Comment by hadibrais** [comment]
@mikedn 
> I suppose it would be nice if `Int32.CompareTo` would read  `m_value`  only once but if you're going to call  `CompareTo`  and modify the  `int`  at the same time without any synchronization then you're going to get burned in many cases no matter what  `CompareTo`  does.

So even if `CompareTo`  was modified to read `m_value` only once, that would still be not good enough because the instance used by `CompareTo` might be the wrong one. So this technique provides some kind of thread-safety but not necessarily the one expected whatever that is.

**Comment by mikedn** [comment]
> So this technique provides some kind of thread-safety but not necessarily the one expected whatever that is.

Yes. Changing `CompareTo` would avoid an obnoxious case - getting a 0 result even if the compared values are never equal. But it doesn't matter what `CompareTo` does, you'll never be able to properly sort an array of `Int32` if another thread keeps changing the values.

**Comment by hadibrais** [comment]
> I think that this is primarily a problem with the documentation's use of the "thread safe"

Agreed. Even across MSDN itself the term might mean different things for different types. Lots of fun.This term should be formalized.
